server/sonar-server/src/main/java/org/sonar/server/platform/ws/HealthActionSupport.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * SonarQube
+ * Copyright (C) 2009-2017 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.platform.ws;
+
+import com.google.common.io.Resources;
+import java.util.Comparator;
+import org.sonar.api.server.ws.WebService;
+import org.sonar.cluster.health.NodeDetails;
+import org.sonar.cluster.health.NodeHealth;
+import org.sonar.server.health.ClusterHealth;
+import org.sonar.server.health.Health;
+import org.sonar.server.health.HealthChecker;
+import org.sonarqube.ws.WsSystem;
+
+public class HealthActionSupport {
+  private static final Comparator<NodeHealth> NODE_HEALTH_COMPARATOR = Comparator.<NodeHealth>comparingInt(s -> s.getDetails().getType().ordinal())
+    .thenComparing(a -> a.getDetails().getName())
+    .thenComparing(a -> a.getDetails().getHost())
+    .thenComparing(a -> a.getDetails().getPort());
+  private final HealthChecker healthChecker;
+
+  public HealthActionSupport(HealthChecker healthChecker) {
+    this.healthChecker = healthChecker;
+  }
+
+  void define(WebService.NewController controller, SystemWsAction handler) {
+    controller.createAction("health")
+      .setDescription("Provide health status of the current SonarQube instance." +
+        "<p>status: the health status" +
+        " <ul>" +
+        " <li>GREEN: SonarQube is fully operational</li>" +
+        " <li>YELLOW: SonarQube is operational but something must be fixed to be fully operational</li>" +
+        " <li>RED: SonarQube is not operational</li>" +
+        " </ul>" +
+        "</p>")
+      .setSince("6.6")
+      .setResponseExample(Resources.getResource(this.getClass(), "example-health.json"))
+      .setHandler(handler);
+  }
+
+  WsSystem.HealthResponse checkNodeHealth() {
+    Health check = healthChecker.checkNode();
+    WsSystem.HealthResponse.Builder responseBuilder = WsSystem.HealthResponse.newBuilder()
```

Comment:
In standalone mode, the response has always an empty `nodes` field:
```
{
  "health": "GREEN",
  "causes": [],
  "nodes": []
}
```
This field does make sense only in cluster mode. To achieve that a dedicated protobuf message should be created (see javadoc of ProtobufJsonFormat):
```
message HealthResponse {
  optional Health health = 1;
  repeated Cause causes = 2;
  optional Nodes nodes = 3;
}

message Nodes {
  repeated Node nodes = 1;
}
```

Reply:
I though it was a problem to have responses with multiple format.

not much worth the effort (and delay), imho, but ok

Reply:
ok